### PR TITLE
[asset-effect-type] `effect_type` argument on `@asset` decorator

### PR DESCRIPTION
**Pull request by sryza**
## Summary & Motivation

This PR explores replacing `@observable_source_asset` with `@asset(effect_type=AssetEffectType.OBSERVE)`. The impetus is the deprecation of the term "source asset".

This is an alternative to [this other PR](https://github.com/dagster-io/dagster/pull/23277) that explores replacing it with `@observable_asset`. I've come around to prefering this approach, because it will allow us to avoid `@observable_` decorators corresponding to `@multi_asset`, `@graph_asset`, and `@graph_multi_asset`.

It sits atop these two PRs that make it less of a mouthful (`execution_type=AssetExecutionType.OBSERVATION` -> `effect_type=AssetEffectType.OBSERVE`):
* https://github.com/dagster-io/dagster/pull/23331
* https://github.com/dagster-io/dagster/pull/23330

Users can use the existing `AutomationCondition` argument on `@asset` to automate their observe computations, now that https://github.com/dagster-io/dagster/pull/23317 is merged.

### CODE REVIEW

**Comment by schrockn** [comment]
Definitely concerned about taking the name `effect` here as I think we should keep that open. However happy with this directly for sure 👍🏻

**Comment by sryza** [comment]
@schrockn I think ideally we'd make a decision here on the name before 1.8. Otherwise we end up in a situation where we've announced that "source assets are deprecated" but still have `@observable_source_asset`, which I believe will cause confusion.

My thinking around `AssetEffectType` was that it could work well with meanings of "effect" that fit with the broader context we've discussed them in. I.e. we'd say something like "The two core orchestration primitives are 'computations' and 'effects'. A computation is something that can be executed. An effect is an effect that a computation has on the world, like materializing an asset, recording a check evaluation, or sending an email".

`AssetEffectType` would also be experimental so we wouldn't be locked in.

That said, not at all tied to "effect type". My prime goal is just to avoid the confusing situation described above.

**Comment by schrockn** [comment]
I'm quite certain that Airlift is going to need to drive the creation of a pure "compute" entity in the asset graph that in my head I am calling an "effect". (Note: I think this will end up being a generalization of asset check as you want the ability to have "blocking" versus "non-blocking" effects) so I am reacting in particular to the word "effect", in addition to the change from execution.

I think "Execution Type" is a good tbh and I still believe we should end up replacing mostly replacing "Materialize" with "Execute".

**Comment by sryza** [comment]
> I think this will end up being a generalization of asset check as you want the ability to have "blocking" versus "non-blocking" effects

Generalizing blocking vs. non-blocking makes a lot of sense to me.

I think an asset check evaluation would fall under the usage of "effect" in this PR.  I.e. a single computation can yield multiple asset check evaluations.

Similarly, imagine a single computation that materializes an asset, sends an email, and also has some third side effect. I imagine that some downstream computation might want to be able to block on the email send, but not care about the third side effect.

**Comment by sryza** [comment]
A related thing to think about here is that I suspect we'll want to enable a single computation to be able to:
- Have different effects on different assets – e.g. observe X then materialize Y
- (More speculative, but hard to rule out in the long run) have different effects on the same asset. Some users have already asked for the ability to have a more general set of asset effects. E.g. migrate column schema, delete. A single computation might carry out multiple of these.

**Comment by schrockn** [comment]
> Have different effects on different assets – e.g. observe X then materialize Y

Yeah I think this ends up leading to a road where the user explicit builds up a graph of compute nodes and then annotates it with metadata about what happens during the compute. The computation then emits a arbitrary set of metadata events (e.g. migrate, delete, etc). This is basically where `AssetsDefinition` is headed.

**Comment by OwenKephart** [comment]
Not to go down too much of a rabbit hole, but I think "blocking" is a pretty weird concept at the moment, and we should avoid proliferating it and instead just rely on the existing dependency system.

IMO, it's more natural for a downstream consumer to say that they "depend on a check" than it is for the upstream producer to say "all downstreams must care about the result of this check". Even just from an internal consistency standpoint, we end up having to translate this blocking / non-blocking information into the dependency layer regardless: https://sourcegraph.com/github.com/dagster-io/dagster/-/blob/python_modules/dagster/dagster/_core/definitions/asset_job.py?L415, and so a) we need to keep track of this blocking/non-blocking distinction deep into the internals of the system and b) lose the flexibility to have different per-edge behavior (i.e. A is blocked by check Foo but B is not).

So long story short is that I see the end state as the ability to construct a graph where there are nodes which represent:

- Some compute function
- Some entity in the graph (asset, check, etc.) that will be altered by this compute function
- Some effect that this compute function will have

A single compute function can map to multiple such nodes (i.e. a single multi-asset function can materialize multiple assets, or a single arbitrary compute function could both migrate and materialize an asset). 

A "blocking asset check" is simply an asset check for which all of the assets downstream of the asset it targets depend on it.

**Comment by sryza** [comment]
Yeah @OwenKephart I also don't have conviction that `blocking` is the right general API.

I was interpreting what @schrockn brought up as that, via some API, downstream computations should be able to have execution dependencies on upstream effects, even when those effects aren't asset materializations.

**Comment by sryza** [comment]
Nick and I chatted in person about naming and settled on "result". This PR is superceded by https://github.com/dagster-io/dagster/pull/23405.
